### ISSUE TITLE
Can't target root with a dynamic tag name

### ISSUE THREAD

**Issue by jaydenseric** [zeit/styled-jsx#147]
You can't [target a root element](https://github.com/zeit/styled-jsx#targeting-the-root) with a [dynamic tag name](http://stackoverflow.com/a/33471928/1596978), using `className`:

```js
export default ({children, level = 2}) => {
  const Heading = `h${level}`
  return (
    <Heading className='root'>
      {children}
      <style jsx>{`
        .root {
          color: red;
        }
      `}</style>
    </Heading>
  )
}
```

**Comment by giuseppeg** [comment]
@jaydenseric interesting use case :)
This is not possible right now because you can't target components at all.

Maybe we can make an exception for component names defined within the scope of the current component.

As for now you can wrap `children` in a `span` and style that one instead.

**Comment by jaydenseric** [comment]
Another use case I encounter is to be able to set a `TextInput` component to either a `textarea` or `input` element via a `multiline` prop. It's not an option to add an inner wrapper for styling; the only workaround is to define two components and return the right one with a conditional.

**Comment by jabacchetta** [comment]
I believe I'm having a similar issue. I'm attempt to use dynamic tag names in combination with `className`.

The following appears to ignore the styles:

```javascript
import React from 'react'

import styles from './styles'

type Props = {
  children: string,
  Element: 'span' | 'p' | 'h1' | 'h2' | 'h3',
  kind: 'smTitle' | 'xlTitle',
}

// class declaration is lost
const TextBroken = ({ children, Element, kind }: Props) => (
  <Element className={kind}>
    {children}
    <style jsx>{styles}</style>
  </Element>
)

// works, but requires a span wrapper
const TextWorking = ({ children, Element, kind }: Props) => (
  <Element>
    <span className={kind}>{children}</span>
    <style jsx>{styles}</style>
  </Element>
)
```

**Comment by simonsmith** [comment]
Bumped into this issue almost immediately when trying to add a className to a `<Link>` component from `react-router` :)

Any thoughts on merging #244?

**Comment by giuseppeg** [comment]
@simonsmith the problem with passing the parent scoped `className` to components is that it may lead to unexpected side effect like descendants being styled due to generic selectors defined in the parent. Eg.

```jsx
// descendant
// className meant to be any className i.e. not added on purpose to pass styled-jsx scoped styles
const Child = ({className, children}) => (
   <div className={'hello ' + className}>{children}</div>
)
```

```jsx
// parent

const Parent = ({children}) => (
   <div>
         {children}
         <Child>hello</Child> {/* didn't plan on passing any className */}
        <style jsx>{`div { margin: 50px  }`}</style> {/* since styled-jsx passes the className to Child the inner one will get the margin too */}
   </div>
)
```

To solve this issue maybe we could allow components selectors like 

```jsx
<style jsx>{`Child { color: red  }`}</style>
```
which generates a unique className for that component, not sure if this would work either.

Also not every 3rd party component is made to accept a `className` prop. It would be cool to find a generic pattern to do both, ideas are welcome!

**Comment by simonsmith** [comment]
@giuseppeg If the className is generated in the parent and passed to the `Child` how does that interfere? Doesn't it end up being like this:

```
.jsx-123456.some-class { /*defined in parent component */

<div class="parent">
  <div class="child jsx-123456 some-class
</div>
```

You may have to school me on how styled-jsx is operating behind the scenes. 

Regardless, any thoughts on how I work around the issue with the `<Link>` component?

**Comment by giuseppeg** [comment]
@simonsmith `jsx-123456` is also applied to the outer `div` (`.parent`).

To style the `Link` you could temporarily use `:global()` until we find a better mousetrap:

```jsx
<div>
   <Link className="my-link">something</Link>
   <style jsx>{`div > :global(.my-link) { color: red }`}</style>
</div>
```

**Comment by arsen** [comment]
styled-jsx is pretty amazing, simple and powerful, except the lack of this super obvious functionality, that anyone would expect from css-in-js library, which is a show stopper.

Do you guys have any plans to support styling custom-defined user components without using :global ?

**Comment by giuseppeg** [comment]
@arsen that's on my radar, probably the next most important feature to add!

**Comment by iamawebgeek** [comment]
@giuseppeg What should I do if I can't wrap my component with a div/span
I have following code:
```js
class Button extends React.Component {
  render() {
    return (
      <React.Fragment>
        <style jsx={true}>{`.button { color: white; }`}</style>
        <button className={`button ${this.props.className}`}>
          {this.props.children}
        </button>
      </React.Fragment>
    )
  }
}

class CoolButton extends React.Component {
  render() {
    return (
      <React.Fragment>
        <style jsx={true}>{`.cool-button { border: 1px solid #808080; }`}</style>
        <Button className={`cool-button ${this.props.className}`}>
          {this.props.children}
        </Button>
      </React.Fragment>
    )
  }
}
```
I just want to extend my component this way, but it is not working, please, suggest some other solution

**Comment by a-ignatov-parc** [comment]
@iamawebgeek this should work

```jsx
class Button extends React.Component {
  render() {
    return (
      <button className={`button ${this.props.className}`}>
        <style jsx>{`.button { color: white; }`}</style>
        {this.props.children}
      </button>
    )
  }
}

class CoolButton extends React.Component {
  render() {
    const scoped = resolveScopedStyles(
      <scope>
        <style jsx>{`.cool-button { border: 1px solid #808080; }`}</style>
      </scope>
    );

    const classNames = [
      scoped.className,
      'cool-button',
      this.props.className,
    ];

    return (
      <React.Fragment>
        <Button className={classNames.join(' ')}>
          {this.props.children}
        </Button>
        {scoped.styles}
      </React.Fragment>
    )
  }
}
```

**Comment by iamawebgeek** [comment]
@a-ignatov-parc Awesome, it worked, thanks. Is there any way to make it more simple like just wrapping everything in `scope` tag?

**Comment by a-ignatov-parc** [comment]
Well, you can modify initial `resolveScopedStyles` helper to something like this

```jsx
function resolveScopedStyles(scope) {
  return ({ children }) => {
    children.props.className += ' ' + scope.props.className;

    return (
      <React.Fragment>
        {children}
        {scope.props.children}
      </React.Fragment>
    )
  };
}
```

And then use it as `className` prefixer for single react element

```jsx
class CoolButton extends React.Component {
  render() {
    const Scoped = resolveScopedStyles(
      <scope>
        <style jsx>{`.cool-button { border: 1px solid #808080; }`}</style>
      </scope>
    );

    return (
      <Scoped>
        <Button className={`cool-button ${this.props.className}`}>
          {this.props.children}
        </Button>
      </Scoped>
    );
  }
}
```

**Comment by giuseppeg** [comment]
When the styles are static it is better to define then outside of the `render` method otherwise they are treated as dynamic.

**Comment by a-ignatov-parc** [comment]
> otherwise they are treated as dynamic.

@giuseppeg I thought that

```jsx
const styles = (
  <style jsx>{`
    .foo { ... }
  `}</style>
);

function Component() {
  return (
    <div>
      {styles}
    </div>
  )
}
```
Is same as 
```jsx
function Component() {
  return (
    <div>
      <style jsx>{`
        .foo { ... }
      `}</style>
    </div>
  )
}
```
Am I wrong? 🤔

Or you talking about?

```jsx
const styles = css`
   .foo { ... }
`;

function Component() {
  return (
    <div>
      <style jsx>{styles}</style>
    </div>
  )
}
```

**Comment by giuseppeg** [comment]
@a-ignatov-parc sorry I made some confusion because I didn't look properly at the examples. Styles that reference variables defined within the scope of the render method are treated as dynamic. Eg.

```jsx

render() {
  const color = 'red'
  return <div>
        {this.props.children} 
         <style jsx>{`div { border: 1px solid ${color}; }`}</style>
    </div>
}
```

If `color` is static just define it outside of the component.

The following syntax is not supported I think:
```jsx
const styles = (
  <style jsx>{`
    .foo { ... }
  `}</style>
);

function Component() {
  return (
    <div>
      {styles}
    </div>
  )
}
```

Wheras the one below (`css`) produces both global and scoped styles:

```jsx
const styles = css`
   .foo { ... }
`;

function Component() {
  return (
    <div>
      <style jsx>{styles}</style>
    </div>
  )
}
```

Somebody needs to work on improving this last behavior https://github.com/zeit/styled-jsx/issues/345

**Comment by a-ignatov-parc** [comment]
@giuseppeg thanks for clarification 👍

**Comment by arsen** [comment]
Any progress on this? :)

**Comment by giuseppeg** [comment]
How would you folks feel about adding a special `styled-jsx` props that will add the scoped classname to components?

```jsx
<Heading className="root" styled-jsx>
      {children}
      <style jsx>{`
        .root {
          color: red;
        }
      `}</style>
</Heading>
```

**Comment by a-ignatov-parc** [comment]
Maybe for base scenarios, we should prefix `className` (if exists) without any additional props? 🤔 

I think most of the developers are expecting such behavior.

**Comment by a-ignatov-parc** [comment]
I thought that we can prefix `className` prop for all react elements (DOM + Components) if it's specified as a string (and maybe some other cases). For advanced scenarios, we can use `css.resolve`.

Also, this behavior can be changed in babel's plugin options. By default, I'd suggest to leave it disabled.

What do you think @giuseppeg?

**Comment by a-ignatov-parc** [comment]
```jsx
<Heading>
  <style jsx>{`
    .root {
      color: red;
    }
  `}</style>
</Heading>
```

Transforms to 

```jsx
<Heading>
  <JSXStyle id="xxx" />
</Heading>
```

---

```jsx
<Heading className="root">
  <style jsx>{`
    .root {
      color: red;
    }
  `}</style>
</Heading>
```

Transforms to 

```jsx
<Heading className="xxx root">
  <JSXStyle id="xxx" />
</Heading>
```
---

But

```jsx
function MyComp({ classes }) {
  return (
    <Heading className={classes}>
      <style jsx>{`
        .root {
          color: red;
        }
      `}</style>
    </Heading>
  );
}
```

Should transform to

```jsx
function MyComp({ classes }) {
  return (
    <Heading className={classes}>
      <JSXStyle id="xxx" />
    </Heading>
  );
}
```

**Comment by a-ignatov-parc** [comment]
I think if `className` prop is static and is a string we can prefix it.

**Comment by giuseppeg** [comment]
I think that it makes sense to automatically scope (add the className) in the case reported by the OP. In any other case we should scope only when the user adds a prop like `styled-jsx` otherwise people can use `css.resolve` i.e. prefer explicitness over magic

**Comment by twltwl** [comment]
Any progress on this ? 😺

**Comment by giuseppeg** [comment]
@twltwl hi, want to pick this up? I can offer guidance

**Comment by twltwl** [comment]
@giuseppeg Yes I could. Some pointers on where to start would be apprichiated :)

**Comment by giuseppeg** [comment]
fixed in #462 will ship with styled-jsx v3
